overlord/restart/restart_test.go

```diff
@@ -395,3 +399,87 @@ func (s *restartSuite) TestPendingForSystemRestart(c *C) {
 	c.Check(rm.PendingForSystemRestart(chg2), Equals, false)
 	c.Check(rm.PendingForSystemRestart(chg3), Equals, false)
 }
+
+type notifyRebootRequiredSuite struct {
+	testutil.BaseTest
+
+	st          *state.State
+	mockNrrPath string
+	mockLog     *bytes.Buffer
+	t1          *state.Task
+}
+
+var _ = Suite(&notifyRebootRequiredSuite{})
+
+func (s *notifyRebootRequiredSuite) SetUpTest(c *C) {
+	s.BaseTest.SetUpTest(c)
+
+	s.AddCleanup(release.MockOnClassic(true))
+
+	s.st = state.New(nil)
+
+	mockLog, restore := logger.MockLogger()
+	s.AddCleanup(restore)
+	s.mockLog = mockLog
+
+	dirs.SetRootDir(c.MkDir())
+	s.mockNrrPath = filepath.Join(dirs.GlobalRootDir, "/usr/share/update-notifier/notify-reboot-required")
+
+	s.st.Lock()
+	defer s.st.Unlock()
+
+	_, err := restart.Manager(s.st, "boot-id-1", nil)
+	c.Assert(err, IsNil)
+
+	// pretend there is a snap that requires a reboot
+	chg1 := s.st.NewChange("not-ready", "...")
+	s.t1 = s.st.NewTask("task", "...")
+	s.t1.Set("reboot-required-snap", "some-snap")
+	chg1.AddTask(s.t1)
+}
+
+func (s *notifyRebootRequiredSuite) TestFinishTaskWithRestartNotifiesRebootRequired(c *C) {
+	s.st.Lock()
+	defer s.st.Unlock()
+
+	mockNrr := testutil.MockCommand(c, s.mockNrrPath, `
+test "$DPKG_MAINTSCRIPT_PACAGE" = "snap:some-snap"
+test "$DPKG_MAINTSCRIPT_NAME" = "postinst"
+`)
+	defer mockNrr.Restore()
+
+	restart.FinishTaskWithRestart(s.t1, state.DoneStatus, restart.RestartSystem, nil)
+	c.Check(mockNrr.Calls(), DeepEquals, [][]string{
+		{"notify-reboot-required"},
+	})
+	c.Check(s.mockLog.String(), Equals, "")
+}
+
+func (s *notifyRebootRequiredSuite) TestFinishTaskWithRestartNotifiesRebootRequiredLogsErr(c *C) {
+	s.st.Lock()
+	defer s.st.Unlock()
+
+	mockNrr := testutil.MockCommand(c, s.mockNrrPath, `echo fail; exit 1`)
+	defer mockNrr.Restore()
+
+	restart.FinishTaskWithRestart(s.t1, state.DoneStatus, restart.RestartSystem, nil)
+	c.Check(mockNrr.Calls(), DeepEquals, [][]string{
+		{"notify-reboot-required"},
+	})
+	c.Check(s.mockLog.String(), Matches, `(?ms).*: cannot notify about pending reboot: fail`)
```

Comment:
I think it's worth checking that a failure to notify doesn't prevent the task from going into a waiting state. We could just check that the function call returns a `state.Wait{}` or even that the task is in `WaitStatus` and has the correct bootID

Reply:
Good point, I updated the test(s).